src/events/parser.rs

```diff
@@ -0,0 +1,280 @@
+use anyhow::Result;
+use serde_json;
+use serde_yaml;
+use std::collections::BTreeMap;
+use std::io::BufReader;
+use std::path::PathBuf;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum ComponentType {
+    actor,
+    capability,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Annotations {
+    version: String,
+    description: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Metadata {
+    name: String,
+    annotations: Annotations,
+}
+
+fn default_link_name() -> Option<String> {
+    Some("default".to_string())
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Properties {
+    image: String,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    contract: Option<String>,
+    #[serde(default = "default_link_name", skip_serializing_if = "Option::is_none")]
+    link_name: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum TraitType {
+    spreadscaler,
+    linkdef,
+}
+#[derive(Debug, Serialize, Deserialize)]
+pub enum SpreadRequirements {
+    app(String),
+    zone(String),
+    ledenabled(bool),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum LinkdefValueItems {
+    uri(String),
+    address(String),
+    port(i32),
+}
+
+fn default_weight() -> Option<i32> {
+    Some(100)
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum SpreadRequirementValues {
+    String(String),
+    bool(bool),
+    i32(i32),
+    char(char),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Spread {
+    name: String,
+    requirements: BTreeMap<String, SpreadRequirementValues>,
+    #[serde(default = "default_weight", skip_serializing_if = "Option::is_none")]
+    weight: Option<i32>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum TraitPropertyItem {
+    LinkdefProperty {
```

Comment:
Could you make each of these be their own structs and then include them here? e.g.

```rust
pub struct SpreadScalerProperty {
        replicas: i32,
        spread: Vec<Spread>,
    }
    
 pub enum TraitPropertyItem {
     SpreadScaler(SpreadScalerProperty)
 }
 ```
 
 The reason being that we will need to use those structs in our future `Scaler` types as config data